Added section on audit and logging

bartlannoeye opened the pull request.

bartlannoeye committed:
Added section on audit and logging

daften commented on Guidelines.md:
```diff
@@ -597,6 +598,51 @@ An API should never trust the input from the client and input SHOULD be validatd
 - Use a secure parser for parsing the incoming messages. If you are using XML, make sure to use a parser that is not vulnerable to XXE and similar attacks.
 
+### 7.14 Logging and audit
+Services compliant with the Digipolis REST API Guidelines MUST write logging to `stdout` and `stderr` according to the [12factor][12factor-logging] guidelines. The format of the logs MUST follow the LogStash JSON standard, one entry per line without zonder `[, ]` or `,`. This is also communicated through the FAST procedure.
```
without zonder? :)

bartlannoeye replied:
copy pasta failure :(

daften commented on Guidelines.md:
```diff
@@ -597,6 +598,51 @@ An API should never trust the input from the client and input SHOULD be validatd
 - Use a secure parser for parsing the incoming messages. If you are using XML, make sure to use a parser that is not vulnerable to XXE and similar attacks.
 
+### 7.14 Logging and audit
+Services compliant with the Digipolis REST API Guidelines MUST write logging to `stdout` and `stderr` according to the [12factor][12factor-logging] guidelines. The format of the logs MUST follow the LogStash JSON standard, one entry per line without zonder `[, ]` or `,`. This is also communicated through the FAST procedure.
```
Do we have to mention the FAST procedure here? Shouldn't this document be included as a reference for API guidelines in every call that goes out, including FAST procedures?

thomasdej replied:
It's also the only place in the guidelines where we mention that a specific guideline is included in the FAST procedure. So I would leave the FAST procedure out of the guidelines. I'm not sure if the complete API guidelines will be referenced in a FAST procedure or just a subset? This should be a discussing point on the next API Workgroup if this wasn't mentioned on the previous API Workgroup meetings. I'll check with Jan if  there is any agreement on this subject.

daften commented on Guidelines.md:
```diff
@@ -597,6 +598,51 @@ An API should never trust the input from the client and input SHOULD be validatd
+| @version         | Logstash format version (e.g. 1) 
+| message          | Formatted log message of the event
+| logger_name      | Name of the logger that logged the event
+| X-Correlation-Id | Correlation ID header of the request
```
X- is usually used in HTTP headers to denote it's a custom header. Do we need that in the name for the field here too?

bartlannoeye replied:
Taken from the FAST procedure spec, asked Jan.

daften replied:
I feel we should check this, because I'm 99% sure this has been decided by non-technical people.

daften commented on Guidelines.md:
```diff
@@ -597,6 +598,51 @@ An API should never trust the input from the client and input SHOULD be validatd
+| @timestamp       | Time of the log event. (yyyy-MM-dd'T'HH:mm:ss.SSSZZ) 
+| @version         | Logstash format version (e.g. 1) 
+| message          | Formatted log message of the event
+| logger_name      | Name of the logger that logged the event
```
Do we have or need a naming convention for loggers?

daften commented on Guidelines.md:
```diff
@@ -597,6 +598,51 @@ An API should never trust the input from the client and input SHOULD be validatd
+    - Progress reporting
+    - End time
+- Audit logs before/after security related events (e.g. impersonation, authentication)
+- Consider logging token validation errors to detect attacks
```
Consider, or should it just be logged?

thomasdej replied:
I guess that the logging of token validation errors should be configured in the API Manager? If so, we should check with Michiel Van Impe if this is possible.

daften replied:
Agreed! @impemi ? ;)

bartlannoeye committed:
Updated comments on logging

bartlannoeye commented on Guidelines.md:
````diff
@@ -597,6 +598,51 @@ An API should never trust the input from the client and input SHOULD be validatd
+
+```json
+{"@timestamp":"2018-02-22T13:23:51.625+00:00","@version":1,"message":"A creation event has been sent to the queue for sms entity with id 38","logger_name":"gent.digipolis.servicefactory.notification.sms.api.module.sms.service.impl.QueueSenderImpl","thread_name":"http-nio-8080-exec-1","level":"INFO","level_value":20000,"X-Correlation-Id":"551aca4a-1b04-45eb-b4f4-e465470ce40b"}
+{"@timestamp":"2018-02-22T13:23:51.628+00:00","@version":1,"message":"Response: [method=POST,uri=/v1/sms,query=<null>,headers={Location=[/v1/sms/38]},body=[locations=[/v1/sms/38]],statusCode=201,responseTime=15]","logger_name":"gent.digipolis.servicefactory.notification.sms.api.module.sms.rest.SmsController","thread_name":"http-nio-8080-exec-1","level":"INFO","level_value":20000,"X-Correlation-Id ...
````
aanpassen naar correlation id

ghijsma replied:
Het is correlationId (lower camel case)

bartlannoeye commented:
Approved door API werkgroep, nog sample aanpassen.

bartlannoeye committed:
Merge branch 'master' into Logging

bartlannoeye committed:
Changed correlation Id

Also removed API manager logging point, can't be done on services itself (@impemi)

bartlannoeye commented:
@ghijsma tabel + sample aangepast. Mag je mergen eens afgecheckt met onze collega's van ELK.

ghijsma reviewed (approved):
Is in accordance with the documentation. Can be merged.
